### PR TITLE
cache client initialization code

### PR DESCRIPTION
**Pull request by unorthodoxgeek**
what I've done:
1. ported the code from memcached-promisify, making a small refactor to only use a single promise function
2. moved the cache client initialization from nodule-graphql into this library

### CODE REVIEW

File: .bumpversion.cfg [hunk 1]
```diff
@@ -0,0 +1,9 @@
+[bumpversion]
+current_version = 0.0.1
```
Comment (jessemyers) [inline .bumpversion.cfg]: Please start at `0.1.0`; we reserve that last number for actual patches (hot fixes)

**Comment by unorthodoxgeek** [comment]
build is failing because tests rely on memcached.

File: src/index.js [hunk 2]
```diff
@@ -0,0 +1,3 @@
+export { default as initializeCacheClient } from './initializeCacheClient';
+
+export default {};
```
Comment (pelegdvir) [inline src/index.js]: ?

File: src/memcached-promisify.js [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+ * @param {string} [config.keyPrefix] - cache key prefix
+ * @param {string} [config.cacheHost] - cache host url
+ * @param {Object} [options] - options passed to memcached
+ *
+ * @example
+ *    new Cache('prefix', 'host');
+ *
+ */
+class Cache {
+    constructor(config, options = { maxExpiration: MAX_EXPIRATION }) {
```
Comment (pelegdvir) [inline src/memcached-promisify.js]: Something is strange about this configuration:
* We have both `config` and `options`
* The default `MAX_EXPIRATION` is const but `localhost:11211` is not

File: src/memcached-promisify.js [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+        this.config = assign({
+            keyPrefix: '',
+            cacheHost: 'localhost:11211',
+        }, config);
+
+        this.config.maxExpiration = options.maxExpiration;
+        console.log('creating memcached instance for cache host:', this.config.cacheHost); // eslint-disable-line
```
Comment (pelegdvir) [inline src/memcached-promisify.js]: if it's already a nodule module, can we replace `console.log` with `logger.debug`

File: src/memcached-promisify.js [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+     * @returns {Promise}
+     */
+    utilGet(key) {
+        return cachePromise(this, 'get', key);
+    }
+
+    /**
+     * get an object of cache items
+     * @param {array} keys - an array of cache keys
+     * @returns {Promise}
```
Comment (pelegdvir) [inline src/memcached-promisify.js]: Can we have more details about the result. Is it an array? Will it return null if it's going to fail, or will it raise an exception.
The same is true to other functions

File: src/memcached-promisify.js [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+     */
+    getMulti(keys) {
+        return cachePromise(this, 'getMulti', keys);
+    }
+
+    /**
+     * gets stats from memcached server
+     * @returns {Promise}
+     */
+    stats() {
```
Comment (pelegdvir) [inline src/memcached-promisify.js]: Do we really need this?
If so - please add detailed comment.
If not - we can delete it.

File: src/__tests__/memcached-promisify.test.js [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+import Cache from '../memcached-promisify';
+
+const cache = new Cache({ keyPrefix: 'test' }, { maxExpiration: 900 });
+describe('memcached promisify', () => {
+    it('should have various methods', () => {
+        expect(cache.get).toBeDefined();
+        expect(cache.utilGet).toBeDefined();
+        expect(cache.getMulti).toBeDefined();
+        expect(cache.set).toBeDefined();
+        expect(cache.add).toBeDefined();
```
Comment (pelegdvir) [inline src/__tests__/memcached-promisify.test.js]: yay

File: src/__tests__/memcached-promisify.test.js [hunk 8]
```diff
@@ -0,0 +1,94 @@
+import { get } from 'lodash';
```
Comment (pelegdvir) [inline src/__tests__/memcached-promisify.test.js]: Can we also have a test with `initializeCacheClient`?

File: src/initializeCacheClient.js [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+    // fix-up caching configuration
+    if (!get(config, 'cache.enabled')) {
+        if (!getMetadata().testing) {
+            terminal.disabled('caching');
+        }
+    } else {
+        if (!getMetadata().testing) {
+            terminal.enabled('caching');
+        }
+        bind('cache', () => createCacheClient());
```
Comment (pelegdvir) [inline src/initializeCacheClient.js]: Should we make the bind string ("cache") configurable?

File: src/initializeCacheClient.js [hunk 10]
```diff
@@ -0,0 +1,35 @@
+/* Memcached cache client. */
+import { get } from 'lodash';
+import { bind, getContainer, getMetadata } from '@globality/nodule-config';
+
+import Cache from './memcached-promisify';
+
+function createCacheClient() {
+    const { config } = getContainer();
+    const memcachedConfig = get(config, 'cache.memcached');
+    const cacheHost = get(config, 'cache.hosts', '').split(',');
```
Comment (pelegdvir) [inline src/initializeCacheClient.js]: Can we use `cache.memcached.cacheHost` instead?

Comment (pelegdvir) [inline src/initializeCacheClient.js]: Do we really have to `.split(',')`, Maybe `memcached` can handle strings

Comment (unorthodoxgeek) [reply src/initializeCacheClient.js]: it'll interfere with the previous line

Comment (unorthodoxgeek) [reply src/initializeCacheClient.js]: it can't. see https://github.com/3rd-Eden/memcached#options

Comment (pelegdvir) [reply src/initializeCacheClient.js]: unless - we don't need two lines

File: src/initializeCacheClient.js (hunk 9 above)
Comment (unorthodoxgeek) [reply src/initializeCacheClient.js]: I don't see why

File: src/initializeCacheClient.js (hunk 10 above)
Comment (unorthodoxgeek) [reply src/initializeCacheClient.js]: moved the split to the Cache constructor, and put the config into cache.memcached

File: src/__tests__/initializeCacheClient.test.js [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+const mockBind = jest.fn();
+const mockDisabled = jest.fn();
+const mockEnabled = jest.fn();
+
+const mockConfig = {
+    config: {
+        cache: {
+            enabled: true,
+            memcached: {
+                hosts: 'localhost:11211,localhost:11212',
```
Comment (pelegdvir) [inline src/__tests__/initializeCacheClient.test.js]: why both?

File: src/__tests__/initializeCacheClient.test.js [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+jest.mock('@globality/nodule-config', () => ({
+    getContainer: () => (mockConfig),
+    getMetadata: () => ({ testing: false }),
+    bind: (key, fn) => mockBind(key, fn),
+}));
+
+describe('initializeCacheClient', () => {
+    it('binds a cache client to the graph', () => {
+        initializeCacheClient();
+        expect(mockBind).toHaveBeenCalledWith('cache', expect.anything());
```
Comment (pelegdvir) [inline src/__tests__/initializeCacheClient.test.js]: Can we have a test for the way that we can access the cache client from the graph?
We like to use tests as a usage example.

File: src/initializeCacheClient.js (hunk 9 above)
Comment (pelegdvir) [reply src/initializeCacheClient.js]: if we're going to use another `cache` service

File: src/initializeCacheClient.js [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+    return client;
+}
+
+/* Configure caching for a nodule-graphql framework.
+ */
+
+export default function initializeCacheClient() {
+    const { config, terminal } = getContainer();
+    // fix-up caching configuration
+    if (!get(config, 'cache.enabled')) {
```
Comment (pelegdvir) [inline src/initializeCacheClient.js]: nit: we can reverse the condition and start with the enabled case

File: src/memcached-promisify.js [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+ * @constructor
+ * @param {string} [cacheHost] - cache host url
+ * @param {Object} [options] - options passed to memcached
+ *
+ * @example
+ *    new Cache('host', { maxExpiration: 900 });
+ *
+ */
+class Cache {
+    constructor(options = {
```
Comment (pelegdvir) [inline src/memcached-promisify.js]: If we're going to pass `{hosts: DEFAULT_HOST}`, what will be the value of `maxExpiration`?

File: src/memcached-promisify.js [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+        maxExpiration: MAX_EXPIRATION,
+        hosts: DEFAULT_HOST,
+    }) {
+        this.maxExpiration = options.maxExpiration;
+        const hosts = options.hosts.split(',');
+        // eslint-disable-next-line no-underscore-dangle
```
Comment (pelegdvir) [inline src/memcached-promisify.js]: nit: can we just rename `this._cache` instead of adding eslint comments?

File: src/initializeCacheClient.js [hunk 16]
```diff
@@ -0,0 +1,33 @@
+/* Memcached cache client. */
+import { get } from 'lodash';
+import { bind, getContainer, getMetadata } from '@globality/nodule-config';
+
+import Cache from './memcached-promisify';
+
+function createCacheClient() {
+    const { config } = getContainer();
+    const memcachedConfig = get(config, 'cache.memcached');
+    const client = new Cache(
```
Comment (saivickna) [inline src/initializeCacheClient.js]: nit client doesnt do anything here so just return new Cache?

File: src/__tests__/initializeCacheClient.test.js (hunk 11 above)
Comment (unorthodoxgeek) [reply src/__tests__/initializeCacheClient.test.js]: why not? we do like to use tests as a usage example...

File: src/initializeCacheClient.js (hunk 9 above)
Comment (unorthodoxgeek) [reply src/initializeCacheClient.js]: we'd put it in the same place. why would we use multiple providers for a single gateway? I don't think there's a reason for solving a problem which doesn't exist...

**Pull request merged**